### PR TITLE
Processed SRM & SSM

### PR DESCRIPTION
**Pull request by iyourshaw**
Adds Processed SRM and SSM messages.

Consumes from the ODE topics:
* topic.OdeSrmJson
* topic.OdeSsmJson

Produces to new topics (see the [related jpo-utils PR](https://github.com/CDOT-CV/jpo-utils/pull/39) that creates the topics)
* topic.ProcessedSrm
* topic.ProcessedSsm

Examples:
* [ProcessedSrm Example](https://github.com/CDOT-CV/jpo-geojsonconverter/blob/3e2e0d523286f26e02ca11fd5a3c60510212167c/jpo-geojsonconverter/src/test/resources/json/sample.processed-srm-multi.json)
* [ProcessedSsm Example](https://github.com/CDOT-CV/jpo-geojsonconverter/blob/3e2e0d523286f26e02ca11fd5a3c60510212167c/jpo-geojsonconverter/src/test/resources/json/sample.processed-ssm-multi.json)

## Desired features of the processed objects:

1. Display as GeoJSON on a map
2. Countability: 1 Processed object per 1 ODE object
3. Ability to match SRMs with SSMs, and to match SSMs with MAPs

## ProcessedSrm Design

SRMs are similar to BSMs in that they are sent from a vehicle and can include the vehicle's location. Unlike BSMs, the location is optional in SRMs, but if it is present it is used as the point geometry of a GeoJSON feature. 

It is possible for one SRM to contain requests for multiple lanes at one intersection, or to target more than one intersection, so `ProcessedSrm` data structures contain a list of requests.  We have observed actual data with multiple requests for different lanes in CDOT's installation. The processed structure retains all the requests in the SRM, even if there happened to be requests for more than one intersection.

The Kafka messages for `ProcessedSrm` use a key containing the RSU ID and VehicleID.  The intersection ID and region are not included in the key because of the possibility of a message targeting more than one intersection.

## ProcessedSsm Design

SSMs are similar to SPATs in that they don't have a geometry of their own but must be matched with a corresponding MAP message to obtain location data.  Therefore, `ProcessedSsm` objects are simple POJOs, not GeoJSON.  

Also, like SPATs, SSMs can contain groups of messages for more than one intersection, but it is anticipated that it is unlikely that they actually would, because they are broadcast from RSUs located at one intersection.  Therefore, this code takes only the first intersection from any SSM it receives, and logs a warning if an SSM contains more than one intersection.  This is the same as how SPATs are treated.

It also would be possible for SSMs to contain responses for more than one vehicle, or for requests for more than one lane for a vehicle.  We have not seen examples of SSMs with multiple statuses in a cursory check of one day of data from CDOT, but this scenario seems more likely that it could happen according to the language in the J2735 where SSMs are defined, which describes them as collections of statuses targeted for multiple vehicles and lanes. For this reason, the `ProcessedSsm` data structure contains a list of statuses.

The Kafka messages for `ProcessedSsm` use RSU ID and Intersection in the key because they are associated with only one intersection.

## CIMMS Considerations

ProcessedSsms are keyed by intersection ID to facilitate joining with MAPs.  ProcessedSrms will need to be rekeyed and repartitioned similarly to how BSMs are handled.  

Since ProcessedSrms and ProcessedSsms contain lists of requests and status, the CIMMS will need to perform a flatMap and rekey operation before joining them together to produce events for each vehicle request, but the CIMMS already does a lot of similar streams manipulations, so that will not be a problem for it.

### CODE REVIEW

**Comment by John-Wiens** [comment]
In the processed SRM, if the SRM is targeting 2 intersections, it is likely that the SRM could be heard by 2 different RSU units as well. If both RSU units receive the SRM message, those 2 entries could be on different partitions because they are received by different RSU units. Is this going to cause any issues downstream with counts? Should we be performing counts based upon individual request ID's?

File: jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/converter/FieldConversions.java [hunk 1]
```diff
@@ -288,4 +301,129 @@ public static ZoneOffset convertDOffset(DOffset dOffset) {
...
+        final ZonedDateTime yearDate = Instant.parse(dateString).atZone(ZoneId.of("UTC"));
+        return yearDate.plusMinutes(moy);
+    }
+
+    public static ZonedDateTime convertMinuteOfYearAndDSecond(final MinuteOfTheYear minuteOfTheYear, final int year, final DSecond dSecond) {
+        ZonedDateTime minuteDate = convertMinuteOfYear(minuteOfTheYear, year);
+        if (minuteDate == null) return null;
+        if (dSecond == null) return minuteDate;
+        SecondNanos secondNanos = convertDSecond(dSecond);
+        if (secondNanos == null) return minuteDate;
```
Comment (John-Wiens) [inline jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/converter/FieldConversions.java]: if dSecond is not null, I think it is not possible for secondNanos to be null. This check might be unnecessary.

File: jpo-geojsonconverter/src/test/java/us/dot/its/jpo/geojsonconverter/validator/SsmJsonValidatorTest.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    @Test
+    public void invalidJsonTest_ByteArray() {
+        testJson_ByteArray(jsonValidator, invalidJsonResource, false);
+    }
+
+    @Test
+    public void testException() {
+        SrmJsonValidator badValidator = new SrmJsonValidator(null);
```
Comment (John-Wiens) [inline jpo-geojsonconverter/src/test/java/us/dot/its/jpo/geojsonconverter/validator/SsmJsonValidatorTest.java]: Should this be SSM?

File: jpo-geojsonconverter/src/test/java/us/dot/its/jpo/geojsonconverter/converter/ssm/SsmTopologyTest.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+import org.apache.kafka.streams.TopologyTestDriver;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
+import org.springframework.core.io.DefaultResourceLoader;
+import org.springframework.core.io.Resource;
+import org.springframework.core.io.ResourceLoader;
+import us.dot.its.jpo.geojsonconverter.converter.srm.SrmConverter;
```
Comment (John-Wiens) [inline jpo-geojsonconverter/src/test/java/us/dot/its/jpo/geojsonconverter/converter/ssm/SsmTopologyTest.java]: Are all of these imports for srm required?

File: jpo-geojsonconverter/src/test/java/us/dot/its/jpo/geojsonconverter/converter/ssm/SsmConverterTest.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    @Parameters
+    public static Collection<Object[]> params() {
+        return Arrays.asList(new Object[][] {
+                { SSM, 1}
+        });
+    }
+
+    public static final String SSM = """
```
Comment (John-Wiens) [inline jpo-geojsonconverter/src/test/java/us/dot/its/jpo/geojsonconverter/converter/ssm/SsmConverterTest.java]: Should this SSM be loaded from a file instead of hardcoded?

File: jpo-geojsonconverter/src/test/java/us/dot/its/jpo/geojsonconverter/converter/srm/SrmConverterTest.java [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+    @Parameters
+    public static Collection<Object[]> params() {
+        return Arrays.asList(new Object[][] {
+                { SRM, 1 },
+                { SRM_2LANES, 2},
+                { SRM_4LANES, 4}
+        });
+    }
+
+    public static final String SRM = """
```
Comment (John-Wiens) [inline jpo-geojsonconverter/src/test/java/us/dot/its/jpo/geojsonconverter/converter/srm/SrmConverterTest.java]: Load SRMs from file?

File: jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/pojos/geojson/srm/ProcessedSignalRequest.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+    // Outbound access point is optional
+    private Integer outboundLaneID;
+    private Integer outboundApproachID;
+    private Integer outboundLaneConnectionID;
+
+    // Timestamp of the DF_SignalRequestPackage
+    private ZonedDateTime estimatedTimeOfArrival;
+
+    private Duration estimatedTimeOfArrivalDurationSeconds;
```
Comment (John-Wiens) [inline jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/pojos/geojson/srm/ProcessedSignalRequest.java:50]: Are any special serialization rules required for Duration to function properly? Does it always serialize to seconds by default?

File: jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/converter/ssm/SsmTopology.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+    public static Topology build(
+            String ssmOdeJsonTopic,
+            String ssmProcessedJsonTopic,
+            SsmJsonValidator ssmJsonValidator,
+            SsmConverter converter) {
+        var builder = new StreamsBuilder();
+
+        builder
+                .stream(ssmOdeJsonTopic,
```
Comment (John-Wiens) [inline jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/converter/ssm/SsmTopology.java:32]: Instead of reading in raw bytes here, would it be better to use a Serdes?

File: jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/converter/ssm/SsmConverter.java [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+import java.util.List;
+
+import static us.dot.its.jpo.geojsonconverter.converter.FieldConversions.*;
+
+@Component
+@Slf4j
+public class SsmConverter {
+
+    public ProcessedSsm processSsm(final SignalStatusMessageMessageFrame ssmFrame) {
+        // We don't know what year it is; assume it is this year.
```
Comment (John-Wiens) [inline jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/converter/ssm/SsmConverter.java]: Do we want to handle the rare edge case where the year rolls over? Perhaps look to see if the minute of year in the message is in December, but the current month is January then select the previous year?

File: jpo-geojsonconverter/pom.xml [hunk 9]
```diff
@@ -17,7 +17,7 @@
 
     <groupId>usdot.jpo.ode</groupId>
     <artifactId>jpo-geojsonconverter</artifactId>
-    <version>3.0.5</version>
+    <version>3.0.6</version>
```
Comment (John-Wiens) [inline jpo-geojsonconverter/pom.xml]: I think this should be a minor version release 3.1.0 since it adds new functionality to the app.

File: jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/converter/FieldConversions.java (hunk 1 above)
Comment (iyourshaw) [reply jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/converter/FieldConversions.java]: true, removed the unnecessary null check

File: jpo-geojsonconverter/src/test/java/us/dot/its/jpo/geojsonconverter/validator/SsmJsonValidatorTest.java (hunk 2 above)
Comment (iyourshaw) [reply jpo-geojsonconverter/src/test/java/us/dot/its/jpo/geojsonconverter/validator/SsmJsonValidatorTest.java]: good catch, fixed it

File: jpo-geojsonconverter/src/test/java/us/dot/its/jpo/geojsonconverter/converter/ssm/SsmTopologyTest.java (hunk 3 above)
Comment (iyourshaw) [reply jpo-geojsonconverter/src/test/java/us/dot/its/jpo/geojsonconverter/converter/ssm/SsmTopologyTest.java]: no, removed unused imports.

File: jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/pojos/geojson/srm/ProcessedSignalRequest.java (hunk 6 above)
Comment (iyourshaw) [reply jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/pojos/geojson/srm/ProcessedSignalRequest.java:50]: Jackson serializes Durations to seconds (as a number with a decimal part) by default.  

It's possible to explicitly specify the ChronoUnits, or to specify units other than seconds in a `@JsonProperty(pattern="UNITS")`  attribute.  But if the units are explicitly specified, Jackson omits the decimal part, so IMO the default behavior is preferable, and the default of seconds is acceptable.

File: jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/converter/ssm/SsmTopology.java (hunk 7 above)
Comment (iyourshaw) [reply jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/converter/ssm/SsmTopology.java:32]: The reason for reading raw bytes is so that we have the raw JSON string to run the validator on, so we can be sure the validator has a "clean" copy of the JSON from the topic, without having to round-trip it through Jackson deserialization-serialization first.

(We're using the [networknt json validation library](https://github.com/networknt/json-schema-validator) )

File: jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/converter/ssm/SsmConverter.java (hunk 8 above)
Comment (iyourshaw) [reply jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/converter/ssm/SsmConverter.java]: Probably a good idea to handle that case, considering there are likely to be more emergency vehicles than average out on New Year's Eve.

File: jpo-geojsonconverter/pom.xml (hunk 9 above)
Comment (iyourshaw) [reply jpo-geojsonconverter/pom.xml]: fixed

File: jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/converter/ssm/SsmTransformer.java [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+    @Override
+    public KeyValue<RsuIntersectionKey, ProcessedSsm> apply(Void rawKey, DeserializedRawMessageFrame rawSsm) {
+
+        try {
+            if (!rawSsm.isValidationFailure()) {
+                OdeMessageFrameData rawValue = rawSsm.getOdeMessageFrameData();
+                OdeMessageFrameMetadata metadata = rawValue.getMetadata();
+                OdeMessageFramePayload payload = rawValue.getPayload();
+                log.info("payload: {}", payload.getData());
```
Comment (drewjj) [inline jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/converter/ssm/SsmTransformer.java]: This log makes this print every SSM message that is processed (that doesn't have validation failures). It is more verbose than the other message types, including SRM.

**Comment by drewjj** [comment]
Can a ProcessedSrm and ProcessedSsm section be added to the README? Looks like we have examples of every other supported message type.

**Comment by drewjj** [comment]
Everything else looks great. Tests pass and it all functions as expected with the test data I tried.

File: jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/converter/ssm/SsmConverter.java (hunk 8 above)
Comment (iyourshaw) [reply jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/converter/ssm/SsmConverter.java]: Handled this and added some unit tests for "minute-of-year" edge cases.

File: jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/converter/ssm/SsmTransformer.java (hunk 10 above)
Comment (iyourshaw) [reply jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/converter/ssm/SsmTransformer.java]: removed this log statement.. It was only for testing during dev

**Comment by iyourshaw** [comment]
> Can a ProcessedSrm and ProcessedSsm section be added to the README? Looks like we have examples of every other supported message type.

Added ProcessedSrm/ProcessedSsm documentation to the README

File: jpo-geojsonconverter/src/test/java/us/dot/its/jpo/geojsonconverter/converter/ssm/SsmConverterTest.java (hunk 4 above)
Comment (iyourshaw) [reply jpo-geojsonconverter/src/test/java/us/dot/its/jpo/geojsonconverter/converter/ssm/SsmConverterTest.java]: addressed

File: jpo-geojsonconverter/src/test/java/us/dot/its/jpo/geojsonconverter/converter/srm/SrmConverterTest.java (hunk 5 above)
Comment (iyourshaw) [reply jpo-geojsonconverter/src/test/java/us/dot/its/jpo/geojsonconverter/converter/srm/SrmConverterTest.java]: addressed

**Review by drewjj — APPROVED** [review]
Looks great!

File: jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/pojos/geojson/srm/ProcessedSignalRequest.java (hunk 6 above)
Comment (John-Wiens) [reply jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/pojos/geojson/srm/ProcessedSignalRequest.java:50]: Perfect. Just wanted to make sure nothing weird was going to happen here.

**Review by John-Wiens — APPROVED** [review]

File: jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/validator/SrmJsonValidator.java [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+/**
+ * JSON validator for SRM messages.
+ */
+@Service
+public class SrmJsonValidator extends AbstractJsonValidator {
+    /**
+     * @param jsonSchemaResource The json schema file in resources/schemas.
+     *                           Injected by Spring DI.
+     */
+    public SrmJsonValidator(@Value("${schema.srm}") Resource jsonSchemaResource) {
```
Comment (Michael7371) [inline jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/validator/SrmJsonValidator.java]: (Blocking): To avoid this being a breaking change for applications that use the GJC as a dependency, can you add a fallback value or error catching to handle this gracefully?

File: jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/validator/SsmJsonValidator.java [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+/**
+ * JSON validator for SSM Messages.
+ */
+@Service
+public class SsmJsonValidator extends AbstractJsonValidator {
+    /**
+     * @param jsonSchemaResource The json schema file in resources/schemas.
+     */
+    public SsmJsonValidator(@Value("${schema.ssm}") Resource jsonSchemaResource) {
+        super(jsonSchemaResource);
```
Comment (Michael7371) [inline jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/validator/SsmJsonValidator.java]: (Blocking): To avoid this being a breaking change for applications that use the GJC as a dependency, can you add a fallback value or error catching to handle this gracefully?

File: README.md [hunk 13]
```diff
@@ -429,7 +429,122 @@ Example `ProcessedRtcm` message:
 }
 ```
 
+### ProcessedSrm
```
Comment (Michael7371) [inline README.md:432]: (Blocking): Not related to the README, but can you update the `SchemaGeneratorUtility` to generate processed schemas for SSM and SRM models and then run this task?

**Review by Michael7371 — CHANGES_REQUESTED** [review]
Looks great! Just a few comments.
- Tests pass.
- SSM/SRM successfully process into processed messages.

File: jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/validator/SrmJsonValidator.java (hunk 11 above)
Comment (iyourshaw) [reply jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/validator/SrmJsonValidator.java]: I don't understand the issue. This PR adds the srm schema in the resources, and includes a default value for the "schema.srm" config in application.yaml, so I don't think there would be any breaking change.

File: jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/validator/SsmJsonValidator.java (hunk 12 above)
Comment (iyourshaw) [reply jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/validator/SsmJsonValidator.java]: I don't understand the issue. This PR adds the schema in the resources, and includes a default value for the config in application.yaml, so I don't think there would be any breaking change.

File: README.md (hunk 13 above)
Comment (iyourshaw) [reply README.md:432]: good catch, generated those

File: jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/validator/SsmJsonValidator.java (hunk 12 above)
Comment (Michael7371) [reply jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/validator/SsmJsonValidator.java]: When applications like the intersection API pull the new version, it will have build errors if the application.yml in the intersection API isn't updated to include this value. Figured I'd bring it up to see if you have any ideas.

Comment (iyourshaw) [reply jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/validator/SsmJsonValidator.java]: Oh interesting, so you mean this happens when pulling it in as a library instead of running it as an independent microservice, I'll see what I can do.

Comment (iyourshaw) [reply jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/validator/SsmJsonValidator.java]: Fixed this by directly loading the schema resources instead of injecting them from the config, since the schema resource is embedded and doesn't need to be configurable.

Verified that all unit test pass in the jpo-cvmanager Intersection API using this fix.

File: jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/validator/SrmJsonValidator.java (hunk 11 above)
Comment (iyourshaw) [reply jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/validator/SrmJsonValidator.java]: Fixed this by directly loading the schema resources instead of injecting them from the config, since the schema resource is embedded and doesn't need to be configurable.

Verified that all unit test pass in the jpo-cvmanager Intersection API using this fix.

**Review by Michael7371 — APPROVED** [review]
LGTM!
